### ISSUE TITLE
Improve StreamBlockValidationError API for use in custom validation methods

### ISSUE THREAD

**Issue by yyken** [wagtail/wagtail#7250]
When validating StreamField using the validators attribute, I always got below exception

 ```
 File "venv/lib/python3.9/site-packages/wagtail/core/blocks/stream_block.py", line 44, in js_args
    error.non_block_errors.as_data(),
AttributeError: 'ValidationError' object has no attribute 'as_data'
```

My validator function like below:

```python
def validate_headings_are_unique(value):
    items = [data[1] for data in value.stream_data if 'heading' in data[0]]
    if len(set(items)) != len(items):
        raise StreamBlockValidationError(
            non_block_errors=ValidationError(
                'All headings must have unique values: %(value)s',
                code='invalid',
                params={'value': items},
            )
        )

```

### Issue Summary

I try to modify **StreamBlockValidationErrorAdapter** class

```python
class StreamBlockValidationErrorAdapter(Adapter):
    js_constructor = 'wagtail.blocks.StreamBlockValidationError'

    def js_args(self, error):
        # change error.non_block_errors.as_data() to error.non_block_errors
        return [error.non_block_errors, {
            block_id: child_errors.as_data()
            for block_id, child_errors in error.block_errors.items()
        }]
```

and change validator function as below:

```python
def validate_headings_are_unique(value):
    items = [data[1] for data in value.stream_data if 'heading' in data[0]]
    if len(set(items)) != len(items):
        raise StreamBlockValidationError(
            block_errors={},
            non_block_errors=[ValidationError(
                'All headings must have unique values: %(value)s',
                code='invalid',
                params={'value': items},
            )]
        )

```

then all work properly.

but I don't know how to test **block_errors**

I hope this helps.

### Technical details

* Python version: 3.9.5
* Django version: 3.2.4
* Wagtail version: 2.13.1
* Browser version: Chrome 91

**Comment by gasman** [comment]
Hi @yyken - thanks for the report.

I don't think this is a bug - the `StreamBlockValidationError` class is designed primarily for use by [the StreamField.clean method](https://github.com/wagtail/wagtail/blob/c6017abca042031fb2f7931b406a4c12e7a9e8a4/wagtail/core/blocks/stream_block.py#L130), and is not documented separately. Based on the code there, `non_block_errors` should be an `ErrorList` object, and `block_errors` should be a dict of `ErrorList` objects - if you change your custom validator method to follow that convention, it should hopefully work.

However, I think we could definitely improve the API of StreamBlockValidationError to make it nicer to use from external code - for example, making it possible to omit the `block_errors` argument. (This was noted separately in https://github.com/wagtail/wagtail/pull/7322#discussion_r666795405 - from the constructor it looks like this is optional, but it really isn't.)

**Comment by PeterDekkers** [comment]
> (This was noted separately in [#7322 (comment)](https://github.com/wagtail/wagtail/pull/7322#discussion_r666795405) - from the constructor it looks like this is optional, but it really isn't.)

This caught me out, too. It'd be amazing if the constructor make these two arguments, `block_errors` and `non_block_errors`, required.

**Comment by gasman** [comment]
Potentially-related issues:

* #3874
* #4122
* #5663
* #7202
* #7997
* #8990
* #9577

Proposed plan of action:

* Add support for non-block errors on StructBlock. There doesn't seem to be a specific issue for this, but this will allow building validation for the common "either link page or link URL must be specified" case without having to craft a StructBlockValidationError that arbitrarily attaches the error to one field or the other.
* Ensure that all blocks that throw their own FooBlockValidationError subclass of ValidationError from `clean` (namely, StructBlock, ListBlock and StreamBlock - any others?) don't assume that they'll get the same type passed back to them at the re-rendering stage, but can also handle plain ValidationErrors. These will be treated the same way as a FooBlockValidationError with only non-block errors. This has two benefits:
  * it hopefully fixes #5663, where a ValidationError is passed in from outside the StreamField's own validation mechanism
  * it means that StructBlock/StreamBlock classes that override `clean` for their own validation, such as the aforementioned "either link page or link URL must be specified" case, and don't care about the mechanism of propagating errors to child blocks, can just throw a plain ValidationError, which is the simplest and most obvious API possible (as in, it's the thing people are likely to try before even looking at the docs).
* Revise ListBlockValidationError to accept a dict of `{block_id: error_list}` as the `block_errors` argument, rather than the weird "list of elements that are either None or error_list, to be paired up with their corresponding block values" setup
* Identify any other developer-friendliness improvements we can make to FooBlockValidationError constructors, such as making the block_errors / non_block_errors kwargs optional - so that they're more suitable for use in custom `clean` methods, rather than only designed for Wagtail's internal use
* Document it!
* Stretch goal: see if it makes sense to ditch Telepath serialisation for ValidationError objects. The client-side objects are just wrappers for `messages` / `blockErrors` / `nonBlockErrors` lists, which makes me think that Telepath is overkill for this and we should just use plain JSON instead.

**Comment by gasman** [comment]
Completed in #10108, #10131, #10144, #10146, #10165.
